integration-tests/test/whitelist.spec.ts

```diff
@@ -0,0 +1,121 @@
+/* Imports: External */
+import { Contract, ContractFactory, Wallet } from 'ethers'
+import { ethers } from 'hardhat'
+import chai, { expect } from 'chai'
+import { solidity } from 'ethereum-waffle'
+import { predeploys } from '@eth-optimism/contracts'
+
+/* Imports: Internal */
+import { OptimismEnv } from './shared/env'
+import { l2Provider } from './shared/utils'
+
+chai.use(solidity)
+
+describe('Whitelist', async () => {
+  const initialAmount = 1000
+  const tokenName = 'OVM Test'
+  const tokenDecimals = 8
+  const TokenSymbol = 'OVM'
+
+  let Factory__ERC20: ContractFactory
+  let env: OptimismEnv
+  before(async () => {
+    env = await OptimismEnv.new()
+    Factory__ERC20 = await ethers.getContractFactory('ERC20', env.l2Wallet)
+  })
+
+  describe('when the whitelist is disabled', () => {
+    it('should be able to deploy a contract', async () => {
+      await expect(
+        l2Provider.send('eth_call', [
+          Factory__ERC20.getDeployTransaction(
+            initialAmount,
+            tokenName,
+            tokenDecimals,
+            TokenSymbol
+          ),
+          'latest',
+          {
+            [predeploys.OVM_DeployerWhitelist]: {
+              state: {
+                ['0x0000000000000000000000000000000000000000000000000000000000000000']:
+                  '0x0000000000000000000000000000000000000000000000000000000000000000',
+              },
+            },
+          },
+        ])
+      ).to.not.be.reverted
+    })
+  })
+
+  describe('when the whitelist is enabled', () => {
+    const sender = '0x' + '22'.repeat(20)
+
+    it('should fail if the user is not whitelisted', async () => {
+      await expect(
+        l2Provider.send('eth_call', [
+          {
+            ...Factory__ERC20.getDeployTransaction(
+              initialAmount,
+              tokenName,
+              tokenDecimals,
+              TokenSymbol
+            ),
+            from: sender,
+          },
+          'latest',
+          {
+            [predeploys.OVM_DeployerWhitelist]: {
+              state: {
+                // Set an owner but don't allow this user to deploy
+                // Owner here is address(1) instead of address(0)
+                ['0x0000000000000000000000000000000000000000000000000000000000000000']:
+                  '0x0000000000000000000000000000000000000000000000000000000000000001',
+              },
+            },
+          },
+        ])
+      ).to.be.reverted
```

Comment:
I would check the exact error message here - very important that developers know that their deployment failed because of whitelist and not just some mysterious revert :)